dask/array/routines.py

```diff
@@ -1012,6 +1008,11 @@ def histogram(a, bins=None, range=None, normed=False, weights=None, density=None
 
     (bins_ref, range_ref), deps = unpack_collections([bins, range])
 
+    if scalar_bins:
+        bins = _linspace_from_delayed(range[0], range[1], bins + 1)
+        # ^ NOTE `range[1]` is safe because of the above check, and the initial check
+        # that range must not be None if `scalar_bins`
+
```

Comment:
You are correct. I moved `_linspace_from_delayed` below `unpack_collections` only.
`unpack_collections` unpacks into `(bins_ref, range_ref)` and it is this `bins_ref` is used to call `_block_hist` further down.
`bins` is now updated from an integer to an array after `bins_ref` has been defined instead of before. The reason I still kept this conversion from integer to array is so that the `bins` return argument keeps returning an array as it did before.